substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSectionImpl.java

```diff
@@ -0,0 +1,396 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.objectfile.elf.dwarf;
+
+import com.oracle.objectfile.BasicProgbitsSectionImpl;
+import com.oracle.objectfile.BuildDependency;
+import com.oracle.objectfile.LayoutDecision;
+import com.oracle.objectfile.LayoutDecisionMap;
+import com.oracle.objectfile.ObjectFile;
+import com.oracle.objectfile.debugentry.ClassEntry;
+import com.oracle.objectfile.elf.ELFObjectFile;
+import org.graalvm.compiler.debug.DebugContext;
+
+import java.nio.ByteOrder;
+import java.util.Map;
+import java.util.Set;
+
+import static com.oracle.objectfile.elf.dwarf.DwarfSections.TEXT_SECTION_NAME;
+
+/**
+ * class from which all DWARF debug sections inherit providing common behaviours.
+ */
+public abstract class DwarfSectionImpl extends BasicProgbitsSectionImpl {
+    protected DwarfSections dwarfSections;
+    public boolean debug = false;
+    public long debugTextBase = 0;
+    public long debugAddress = 0;
+    public int debugBase = 0;
+
+    public DwarfSectionImpl(DwarfSections dwarfSections) {
+        this.dwarfSections = dwarfSections;
+    }
+
+    /**
+     * creates the target byte[] array used to define the section contents.
+     *
+     * the main task of this method is to precompute the size of the debug section. given the
+     * complexity of the data layouts that invariably requires performing a dummy write of the
+     * contents, inserting bytes into a small, scratch buffer only when absolutely necessary.
+     * subclasses may also cache some information for use when writing the contents.
+     */
+    public abstract void createContent();
+
+    /**
+     * populates the byte[] array used to contain the section contents.
+     *
+     * in most cases this task reruns the operations performed under createContent but this time
+     * actually writing data to the target byte[].
+     */
+    public abstract void writeContent(DebugContext debugContext);
+
+    @Override
+    public boolean isLoadable() {
+        /*
+         * even though we're a progbits section impl we're not actually loadable
+         */
+        return false;
+    }
+
+    public String debugSectionLogName() {
```

Comment:
Since we will also have Windows debug info generation it might be better to use
```
return "debug.dwarf." + getSectionName();
```
here so that we can have
```
return "debug.pdb." + getSectionName();
```
for Windows.

Reply:
Sure will correct that. I guess I should also leave the section name in lower case as well.

Reply:
> I guess I should also leave the section name in lower case as well.

Yes. That would fit better to our existing use of scope names.